torch/utils/_sympy/solve.py

```diff
@@ -0,0 +1,120 @@
+    # Give up if 'expr' is not a relational operation or if 'thing'
+    # is not on the left-hand side.
+    if not isinstance(expr, sympy.Rel) or not expr.lhs.has(thing):
+        return expr
+
+    lhs, rhs = expr.args
+
+    # Move any constants in the left-hand side to the right-hand side.
+    lhs_const = (
+        sum([a for a in lhs.args if isinstance(a, sympy.Integer)])
+        if isinstance(lhs, sympy.Add)
+        else 0
+    )
+    lhs = lhs - lhs_const  # type: ignore[arg-type]
+    rhs = rhs - lhs_const  # type: ignore[arg-type]
+
+    # Divide both sides by the factors that don't contain thing.
+    if isinstance(lhs, sympy.Mul):
+        other = sympy.Mul(*[a for a in lhs.args if not a.has(thing)])
+        # TODO: mirror the operation if 'other' is negative.
+        if isinstance(expr, (sympy.Eq, sympy.Ne)) or other.is_positive:
+            lhs = lhs / other
+            rhs = rhs / other
+
+    ################################################################################
+    # left-hand side is FloorDiv
+    ################################################################################
+    #
+    # Given the expression: a // b op c
+    # where 'op' is a relational operation, these rules only work if:
+    #   - b > 0
+    #   - c is an integer
+    if isinstance(lhs, FloorDiv) and lhs.divisor.is_positive and rhs.is_integer:
+        # a // b == expr
+        # => a >= (b * expr) and a < (b * (expr + 1))
+        if isinstance(expr, sympy.Eq):
+            numerator, denominator = lhs.args
+            return sympy.And(
+                sympy.Ge(numerator, (rhs * denominator)),  # type: ignore[arg-type]
+                sympy.Lt(numerator, ((rhs + 1) * denominator)),  # type: ignore[arg-type]
+            )
+        # a // b != expr
+        # => a < (b * expr) or a >= (b * (expr + 1))
+        if isinstance(expr, sympy.Ne):
+            numerator, denominator = lhs.args
+            return sympy.Or(
+                sympy.Lt(numerator, (rhs * denominator)),  # type: ignore[arg-type]
+                sympy.Ge(numerator, ((rhs + 1) * denominator)),  # type: ignore[arg-type]
+            )
+        # The transformations below only work if b is positive.
+        # Note: we only have this information for constants.
+        # a // b > expr  => a >= b * (expr + 1)
+        # a // b >= expr => a >= b * expr
+        if isinstance(expr, (sympy.Gt, sympy.Ge)):
+            quotient = rhs if isinstance(expr, sympy.Ge) else (rhs + 1)  # type: ignore[arg-type]
+            return sympy.Ge(lhs.args[0], (quotient * lhs.args[1]))  # type: ignore[arg-type]
+        # a // b < expr  => a < b * expr
+        # a // b <= expr => a < b * (expr + 1)
+        if isinstance(expr, (sympy.Lt, sympy.Le)):
+            quotient = rhs if isinstance(expr, sympy.Lt) else (rhs + 1)  # type: ignore[arg-type]
+            return sympy.Lt(lhs.args[0], (quotient * lhs.args[1]))  # type: ignore[arg-type]
```

Comment:
You have fairly similar logic at https://github.com/pytorch/pytorch/blob/d54cf0254328a56bb5615465e7f5ea5e1799d1cd/torch/fx/experimental/symbolic_shapes.py#L3010-L3026

Is the duplication intentional?

Reply:
No. Forgot to delete that.